pkg/api/query/v1.go

```diff
@@ -1037,6 +1049,48 @@ func (qapi *QueryAPI) stores(_ *http.Request) (interface{}, []error, *api.ApiErr
 	return statuses, nil, nil, func() {}
 }
 
+func (qapi *QueryAPI) getLabelMatchers(matchers []string, tenant string) ([][]*labels.Matcher, *api.ApiError) {
+	tenantLabelMatcher := &labels.Matcher{
+		Name:  qapi.tenantLabel,
+		Type:  labels.MatchEqual,
+		Value: tenant,
+	}
+
+	var matcherSets [][]*labels.Matcher
+
+	// If tenancy is enforced, but there are no matchers at all, add the tenant matcher
+	if len(matchers) == 0 && qapi.enforceTenancy {
+		var matcher []*labels.Matcher
+		matcher = append(matcher, tenantLabelMatcher)
+		matcherSets = append(matcherSets, matcher)
+		return matcherSets, nil
+	}
+
+	for _, s := range matchers {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return nil, &api.ApiError{Typ: api.ErrorBadData, Err: err}
+		}
+		if qapi.enforceTenancy {
+			// first check if there's a tenant matcher already, in which case we overwrite it
+			found := false
+			for idx, matchValue := range matchers {
+				if matchValue.Name == qapi.tenantLabel {
+					matchers[idx] = tenantLabelMatcher
+					found = true
```

Comment:
Couldn't we break here?

Reply:
Oh ok. Maybe then add a comment explaining that. I am sure someone will add a break otherwise 😄

Reply:
As an optimization, I'd suggest only keeping one tenant matcher and deleting others.

Reply:
That makes sense, pushed a fix for that :+1: